flang/docs/tutorials/addingIntrinsics.md

```diff
@@ -0,0 +1,383 @@
+
+In `lib/Evaluate/fold-character.cpp`, there is case to fold trim that looks like:
+```
+  } else if (name == "trim") { // not elemental
+    if (auto scalar{
+            GetScalarConstantArguments<T>(context, funcRef.arguments())}) {
+      return Expr<T>{Constant<T>{
+          CharacterUtils<KIND>::TRIM(std::get<Scalar<T>>(*scalar))}};
+    }
+  }
+```
+It uses a helper function operating on the `Scalar` abstract representation. `Scalar<T>` when T is a character kind 1
+are simply `std::string`, so `CharacterUtils<1>::TRIM` actually simply implements Trim on a `std::string`.
+The folding code above mainly try to extract a constant from the argument, and if successful, repackage the result
+of the helper in a new typed expression representation (`Expr<T>`).
+### TRIM runtime
+Trim runtime API is defined in `runtime/character.h` as:
+
+```
+void RTNAME(Trim)(Descriptor &result, const Descriptor &string,
+    const char *sourceFile = nullptr, int sourceLine = 0);
+```
+
+The `RTNAME` macro adds some mangling to the runtime to avoid collision with user symbols and handle
+runtime library versioning. Here, it only matters to know it must be used in all function declaration of
+the runtime API.
+
+Trim runtime takes a descriptor for the result that it will allocate, and a constant descriptor for the string argument.
+On top of that, it takes optional source file and line arguments, that are here to print source location if the runtime
+was to reach a critical error (e.g. if the provided descriptor is actually not describing a character scalar).
+
+It is rather simple and follows the standard definition of Trim in the argument naming. In general, it is best to stick
+to the intrinsic interface in terms of argument names and order if possible.
+
+The actual implementation is in `runtime/character.cpp` and is not copied here. Have a look at it. A notable point is that it
+is not using `std::string` anywhere, because that would bring the C++ runtime as a dependency.
+
+It is using some of the `Descriptor` member function such as `string.OffsetElement<>`,
+`result.Establish()`, `result.Allocate()`, or `string.ElementBytes()`. It is worth having a look at
+`flang/runtime/descriptor.h` to understand more about the descriptor format and the tools available for it.
+
+The error handling in the runtime is done with the `Terminator` class. See the `terminator.Crash()`
+and `RUNTIME_CHECK` macro. `terminator.Crash` causes an unconditional crash, while `RUNTIME_CHECK`
+will cause a crash if the boolean provided to it is not true. The runtime should always crash rather
+than reach undefined states.
+
+Last notable point is that it is using LEN_TRIM implementation rather than to re-implement it locally.
+It is always a good idea to share part of implementations that are similar between related intrinsics.
+It is OK to define internal helper functions, classes or to use templates to achieve this.
+
+
+### TRIM Lowering
```

Comment:
It would be good to have an example of testing TRIM lowering.

Reply:
I have added a lowering regression test example in in https://github.com/flang-compiler/f18-llvm-project/commit/11b8022d47badec9430f126523630e4043969416